### PR TITLE
Add forgot and reset password endpoints

### PR DESCRIPTION
**Pull request by caiseL**

### CODE REVIEW

File: lib/api/shared/src/main/scala/net/wiringbits/api/models/ResetPassword.scala [hunk 1]
```diff
@@ -0,0 +1,13 @@
+package net.wiringbits.api.models
+
+import net.wiringbits.common.models.{Password, UserToken}
+import play.api.libs.json.{Format, Json}
+
+object ResetPassword {
+  case class Request(token: UserToken, password: Password)
+  case class Response(token: Option[String])
```
Comment (AlexITC) [inline lib/api/shared/src/main/scala/net/wiringbits/api/models/ResetPassword.scala:8]: When would we get a `None`? I think there is no such case

File: server/src/main/scala/controllers/UsersController.scala [hunk 2]
```diff
@@ -41,6 +41,22 @@ class UsersController @Inject() (
...
+    for {
+      response <- usersService.forgotPassword(body)
+    } yield Ok(Json.toJson(response))
+  }
+
+  def resetPassword() = handleJsonBody[ResetPassword.Request] { request =>
+    val body = request.body
+    logger.info(s"Reset user's password: ${body.token.userId}")
+    for {
+      response <- usersService.resetPassword(body.token.userId, body.password)
```
Comment (AlexITC) [inline server/src/main/scala/controllers/UsersController.scala:56]: This is wrong, we need to verify the token is correct instead of just trusting it

File: server/src/main/scala/net/wiringbits/repositories/daos/UsersDAO.scala [hunk 3]
```diff
@@ -64,6 +64,14 @@ object UsersDAO {
     """.execute()
   }
 
+  def resetPassword(userId: UUID, password: String)(implicit conn: Connection): Unit = {
+    val _ = SQL"""
+      UPDATE users
+      SET password = $password
```
Comment (AlexITC) [inline server/src/main/scala/net/wiringbits/repositories/daos/UsersDAO.scala:70]: I think we have an `updated_at` column that must be updated too

File: server/src/main/scala/net/wiringbits/services/UsersService.scala [hunk 4]
```diff
@@ -83,6 +89,35 @@ class UsersService @Inject() (
     } yield Login.Response(user.id, user.name, user.email, token)
   }
 
+  def forgotPassword(request: ForgotPassword.Request): Future[ForgotPassword.Response] = {
+    for {
+      _ <- validateCaptcha(request.captcha)
+      userMaybe <- repository.find(request.email)
+      _ = userMaybe.foreach(user => {
```
Comment (AlexITC) [inline server/src/main/scala/net/wiringbits/services/UsersService.scala:96]: `.foreach { user => ... }` works, still, this is bad because you are discarding the result while we care if the email was sent, you can do this:

```
user = userMaybe.getOrElse(throw new RuntimeException("Email not registered"))
_ <- enforeVerifiedUser(user)
_ <- emailApi.sendEmail(...)
```

It is ideal to not fail when the user isn't registered but mention something like `If your account is registered you should get an email to reset your password`, which we can do in another PR

File: server/src/main/scala/net/wiringbits/services/UsersService.scala [hunk 5]
```diff
@@ -83,6 +89,35 @@ class UsersService @Inject() (
...
+        if (user.verifiedOn.isEmpty)
+          throw new RuntimeException(s"User's email ${request.email} hasn't been verified yet")
+        val emailMessage = EmailMessage.forgotPassword(user.name, webAppConfig.host, s"${user.id}")
+        emailApi.sendEmail(EmailRequest(user.email, emailMessage))
+      })
+    } yield ForgotPassword.Response()
+  }
+
+  def resetPassword(userId: UUID, password: Password): Future[ResetPassword.Response] = {
```
Comment (AlexITC) [inline server/src/main/scala/net/wiringbits/services/UsersService.scala:105]: We need to validate the token

File: server/src/main/scala/net/wiringbits/services/UsersService.scala [hunk 6]
```diff
@@ -83,6 +89,35 @@ class UsersService @Inject() (
...
+    val userMaybe = repository.find(userId)
+    userMaybe.flatMap {
+      case None => Future(ResetPassword.Response(None))
```
Comment (AlexITC) [inline server/src/main/scala/net/wiringbits/services/UsersService.scala:108]: If the token exists the user must exists too

File: server/src/main/scala/net/wiringbits/services/UsersService.scala [hunk 7]
```diff
@@ -83,6 +89,35 @@ class UsersService @Inject() (
...
+      case Some(user) =>
+        val hashedPassword = BCrypt.hashpw(password.string, BCrypt.gensalt())
+        for {
+          _ <- repository.resetPassword(userId, hashedPassword)
+          _ <- userLogsRepository.create(user.id, "Password was reset")
```
Comment (AlexITC) [inline server/src/main/scala/net/wiringbits/services/UsersService.scala:113]: The logs need to be created in the transaction where the password is updated so that both operations are done atomically

File: server/src/main/scala/net/wiringbits/util/EmailMessage.scala [hunk 8]
```diff
@@ -24,4 +24,28 @@ object EmailMessage {
...
+         |If you did not perform this request, you can safely ignore this email.
+         |""".stripMargin
+
+    EmailMessage(subject, body)
+  }
+  def resetPassword(name: Name): EmailMessage = {
+    val subject = "Your SWAN password has been reset"
+    val body =
+      s"""Hi ${name.string},
+         |<h2>Your SWAN password has been changed.</h2>
```
Comment (AlexITC) [inline server/src/main/scala/net/wiringbits/util/EmailMessage.scala:45]: Fix this text

File: server/src/test/scala/controllers/UsersControllerSpec.scala [hunk 9]
```diff
@@ -284,4 +285,203 @@ class UsersControllerSpec extends PlayPostgresSpec with LoginUtils {
...
+
+      createVerifyLoginUser(request, client).futureValue
+
+      val forgotPasswordRequest = ForgotPassword.Request(email, Captcha.trusted("test"))
+      val response = client.forgotPassword(forgotPasswordRequest).futureValue
+
+      response must be(ForgotPassword.Response())
+    }
+
+    "ignore the request when the user tries to reset a password for unknown email" in withApiClient { client =>
```
Comment (AlexITC) [inline server/src/test/scala/controllers/UsersControllerSpec.scala:308]: This is unnecessary, we need to ignore the request when the user clicks forgets its password but such an email doesn't exists.

File: server/src/test/scala/controllers/UsersControllerSpec.scala [hunk 10]
```diff
@@ -284,4 +285,203 @@ class UsersControllerSpec extends PlayPostgresSpec with LoginUtils {
...
+        ResetPassword.Request(UserToken(userId), Password.trusted("test456..."))
+
+      val response = client
+        .resetPassword(resetPasswordRequest)
+        .futureValue
+
+      response mustNot be(ResetPassword.Response(None))
+    }
+
+    "return a None token when a nonexistent user tries to reset a password" in withApiClient { client =>
```
Comment (AlexITC) [inline server/src/test/scala/controllers/UsersControllerSpec.scala:438]: This is not required, we must fail if this ever happens, if the token is valid the user must exist

**Review by AlexITC — CHANGES_REQUESTED** [review]

File: server/src/main/scala/net/wiringbits/repositories/daos/UsersDAO.scala (hunk 3 above)
Comment (caiseL) [reply server/src/main/scala/net/wiringbits/repositories/daos/UsersDAO.scala:70]: It seems like we don't have it. Should I add it?

File: server/src/main/scala/controllers/UsersController.scala (hunk 2 above)
Comment (caiseL) [reply server/src/main/scala/controllers/UsersController.scala:56]: I don't get it. Do you mean we need tokens for this approach? That's what I was planning to do after #114 gets merged

File: server/src/test/scala/controllers/UsersControllerSpec.scala (hunk 9 above)
Comment (AlexITC) [reply server/src/test/scala/controllers/UsersControllerSpec.scala:308]: Actually, I got confused, this is ok.

File: server/src/main/scala/net/wiringbits/config/UserTokensConfig.scala [hunk 11]
```diff
@@ -4,13 +4,14 @@ import play.api.Configuration
...
+case class UserTokensConfig(emailVerificationExp: FiniteDuration, resetPasswordExp: FiniteDuration)
 
 object UserTokensConfig {
 
   def apply(conf: Configuration): UserTokensConfig = {
     val emailVerificationExp = conf.get[FiniteDuration]("emailVerification.expirationTime")
+    val resetPasswordExp = conf.get[FiniteDuration]("resetPassword.expirationTime")
 
-    UserTokensConfig(emailVerificationExp)
+    UserTokensConfig(emailVerificationExp, resetPasswordExp)
```
Comment (AlexITC) [inline server/src/main/scala/net/wiringbits/config/UserTokensConfig.scala:15]: Use named arguments.

**Review by AlexITC — COMMENTED** [review]
Looks good!

Now, for the next PR let's use HMAC for storing the user tokens.

Thanks.

**Review by AlexITC — APPROVED** [review]
